### PR TITLE
[Feature][EPLB] Add asynchronous EPLB support for Model Runner V2

### PR DESCRIPTION
**Pull request by freyfwt**
### What this PR does / why we need it?

This PR completes the asynchronous Model Runner V2 EPLB follow-up to RFC #13410. It keeps vLLM's EPLB controller, load-window lifecycle, default placement policy, placement state, and asynchronous worker as the source of truth, while adding the Ascend adapters required for graph-safe load collection, staged expert movement, and routing refresh.

This PR deliberately does not add or override an EPLB placement policy. Policy selection and execution use the upstream `default` implementation.

Refs #13410

Implementation and review hotspots:

- Organize the shared Ascend EPLB state and communicator under `vllm_ascend.distributed.eplb`, preserving the intended directory refactor.
- Maintain a graph-stable replica routing table and map logical expert IDs to local physical experts in a graph-safe device operation. Physical-expert load recording excludes padding and follows the configured `load_collection_phase`.
- Preserve quantization-owned expert weight views so asynchronous movement updates the weights used by the active fused-MoE implementation.
- Adapt upstream asynchronous transfers to Gloo CPU staging on Ascend. After a layer is committed, refresh its Ascend routing table before acknowledging the upstream worker result.
- Require asynchronous mode and the Gloo communicator for MRv2 EPLB, and reject elastic EP because that combination is not supported.
- Keep upstream release-contract compatibility isolated in the downstream state/controller adapters.

### Does this PR introduce _any_ user-facing change?

Yes. Model Runner V2 EPLB is available on Ascend with asynchronous expert movement and the upstream default placement policy. Users enable it with `--enable-eplb`. Ascend selects `torch_gloo` when the communicator is unset, normalizes MRv2 EPLB to asynchronous mode, and supports `additional_config.eplb_config.load_collection_phase` with `all`, `prefill`, or `decode`.

Elastic EP and non-Gloo communicators are rejected for this path. Legacy Model Runner V1 EPLB and EPLB-disabled execution are unchanged.

### How was this patch tested?

- `bash format.sh ci`: all repository hooks passed.
- Full mypy passed for Python 3.10, 3.11, and 3.12 across `vllm_ascend`, `examples`, and `tests` in a remote Linux lint environment.
- Focused unit tests: 52 passed, covering EPLB state, communicator staging, routing/load operations, configuration validation, ordered async acknowledgement, fused-MoE integration, and load-collection phase handling.
- Runtime-stack preflight passed with the pinned vLLM source and required Ascend custom-op symbol available.
- Real Ascend NPU routing/load operator test: 1 passed.
- Real four-card Ascend E2E: 1 passed with DP2 × TP2, EP world size four, W8A8, asynchronous scheduling, `FULL_AND_PIECEWISE`, prefill-only load collection, and four redundant experts. The test verified answer prefixes and observed a complete asynchronous EPLB cycle.

### Documentation

- Updated the Model Runner V2 EPLB architecture to describe upstream ownership, graph-safe load recording, Gloo staging, and commit ordering.
- Updated the EPLB user guide, additional configuration reference, support matrix, localization files, and Model Runner V1 architecture cross-links.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/worker/v2/model_runner.py [hunk 1]
```diff
@@ -385,6 +386,13 @@ def prepare_inputs(
         is_prefilling_np = num_computed_prefill_tokens_np < prefill_len_np
         batch_has_prefill = bool(np.any(is_prefilling_np))
         self.eplb.set_batch_phase(batch_has_prefill)
+        tp_size = self.parallel_config.tensor_parallel_size
+        tp_rank = get_tensor_model_parallel_rank()
+        quotient, remainder = divmod(num_tokens, tp_size)
+        self.eplb.prepare_forward(
+            self.model_config,
+            quotient + int(tp_rank < remainder),
+        )
```
[the remark this answers is not in the document]
Comment (freyfwt) [reply vllm_ascend/worker/v2/model_runner.py]: Thanks. This code path is no longer part of the PR: the current diff does not modify `model_runner.py`. Load-phase filtering now lives in the EPLB controller/state path, with focused tests for prefill/decode filtering and rank-local step alignment. Resolving this as outdated.

File: vllm_ascend/distributed/eplb/async_worker.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            with torch.cuda.stream(cuda_stream):
+                old_mapping = model_state.physical_to_logical_map.cpu()
+            new_mapping = _run_rebalance_experts(model_state, state, old_mapping, cuda_stream)
+            changed_layers: list[int] = (
+                torch.any(new_mapping != old_mapping, dim=1).nonzero(as_tuple=False).flatten().tolist()
+            )
+
+            for layer_idx in changed_layers:
+                flag = torch.tensor([int(model_state.rebalanced)], dtype=torch.int32, device="cpu")
+                torch.distributed.all_reduce(flag, group=eplb_cpu_group)
```
[the remark this answers is not in the document]
Comment (freyfwt) [reply vllm_ascend/distributed/eplb/async_worker.py]: Thanks for flagging the collective ordering risk. The two coordination paths use distinct process groups: foreground freshness/result-readiness uses `get_ep_group().cpu_group`, while the background worker uses `get_eplb_group().cpu_group`. Gloo-staged weight movement also runs on that background thread before it publishes the pending result; the main thread only commits the staged buffer and records the consumed event. The four-card DP2 × TP2 E2E completed a 25-layer asynchronous transfer cycle without a collective mismatch. No additional process group is needed here.

**Comment by freyfwt** [comment]
Technical preflight is complete at `6355ed03c`: repository formatting passed, full mypy passed for Python 3.10/3.11/3.12, 128 focused CPU tests passed, the real NPU operator test passed, and the four-card DP2 × TP2 E2E passed with a committed 25-layer asynchronous transfer cycle.

Could a maintainer please add the `mrv2` and `ready-precise` labels? The PR author account cannot apply repository labels, and the labels are required to schedule the selected hardware tests and satisfy `ci-gate`.

File: vllm_ascend/distributed/eplb/async_worker.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    assert state.is_async
+
+    def thread_target() -> None:
+        assert device_index is not None
+        torch.accelerator.set_device_index(device_index)
+        stream = torch.cuda.Stream(device=device_index)
+        try:
+            transfer_run_periodically(state=state, cuda_stream=stream, is_profile=is_profile)
+        except Exception as exc:  # pragma: no cover - diagnostic path
+            logger.exception("async loop error (Rank %d): %s", rank, str(exc))
```
Comment (Spicy-Stick) [inline vllm_ascend/distributed/eplb/async_worker.py]: logger.exception() only logs the error; the worker thread then exits while rebalanced may remain True. The main thread can consequently wait forever for a pending_result, and drain_async() may never complete. Please propagate the failure to the main thread or coordinate cleanup across ranks

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- Doc Test: still in progress, retry /rerun after completion
- E2E: still in progress, retry /rerun after completion

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Failed:**
- Doc Test: still in progress, retry /rerun after completion

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test
- E2E

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test

**Comment by freyfwt** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- Doc Test

**Review by weijinqian0 — APPROVED** [review]
